src/utils/proxy.js

```diff
@@ -0,0 +1,346 @@
+    }
+    const urlForAddons = addonUrl(options.addonUrls, req)
+    if (urlForAddons) {
+      return proxy.web(req, res, { target: urlForAddons })
+    }
+
+    return proxy.web(req, res, { ...options, status })
+  }
+
+  return proxy.web(req, res, options)
+}
+
+function initializeProxy(port, distDir, projectDir) {
+  const proxy = httpProxy.createProxyServer({
+    selfHandleResponse: true,
+    target: {
+      host: 'localhost',
+      port,
+    },
+  })
+
+  const headersFiles = Array.from(new Set([path.resolve(projectDir, '_headers'), path.resolve(distDir, '_headers')]))
+
+  let headerRules = headersFiles.reduce((prev, curr) => Object.assign(prev, parseHeadersFile(curr)), {})
+  onChanges(headersFiles, () => {
+    console.log(
+      `${NETLIFYDEVLOG} Reloading headers files`,
+      headersFiles.filter(fs.existsSync).map(p => path.relative(projectDir, p))
+    )
+    headerRules = headersFiles.reduce((prev, curr) => Object.assign(prev, parseHeadersFile(curr)), {})
+  })
+
+  proxy.on('error', err => console.error('error while proxying request:', err.message))
+  proxy.on('proxyReq', (proxyReq, req) => {
+    if (req.originalBody) {
+      proxyReq.write(req.originalBody)
+    }
+  })
+  proxy.on('proxyRes', (proxyRes, req, res) => {
+    if (proxyRes.statusCode === 404 || proxyRes.statusCode === 403) {
+      if (req.alternativePaths && req.alternativePaths.length > 0) {
+        req.url = req.alternativePaths.shift()
+        return proxy.web(req, res, req.proxyOptions)
+      }
+      if (req.proxyOptions && req.proxyOptions.match) {
+        return serveRedirect(req, res, handlers, req.proxyOptions.match, req.proxyOptions)
+      }
+    }
+    const requestURL = new url.URL(req.url, `http://${req.headers.host || 'localhost'}`)
+    const pathHeaderRules = objectForPath(headerRules, requestURL.pathname)
+    if (!isEmpty(pathHeaderRules)) {
+      Object.entries(pathHeaderRules).forEach(([key, val]) => res.setHeader(key, val))
+    }
+    res.writeHead(req.proxyOptions.status || proxyRes.statusCode, proxyRes.headers)
+    proxyRes.on('data', function(data) {
+      res.write(data)
+    })
+    proxyRes.on('end', function() {
+      res.end()
+    })
+  })
+
+  const handlers = {
+    web: (req, res, options) => {
+      const requestURL = new url.URL(req.url, 'http://localhost')
+      req.proxyOptions = options
+      req.alternativePaths = alternativePathsFor(requestURL.pathname).map(p => p + requestURL.search)
+      // Ref: https://nodejs.org/api/net.html#net_socket_remoteaddress
+      req.headers['x-forwarded-for'] = req.connection.remoteAddress || ''
+      return proxy.web(req, res, options)
+    },
+    ws: (req, socket, head) => proxy.ws(req, socket, head),
+  }
+
+  return handlers
+}
+
+async function startProxy(settings = {}, addonUrls, configPath, projectDir) {
+  const functionsServer = settings.functionsPort ? `http://localhost:${settings.functionsPort}` : null
```

Comment:
The original `startProxy` had some code to wait for the framework server port and functions server port:
https://github.com/netlify/cli/blob/93468a3fc43db5d637b30b9cbb21042da8626962/src/commands/dev/index.js#L170
https://github.com/netlify/cli/blob/93468a3fc43db5d637b30b9cbb21042da8626962/src/commands/dev/index.js#L178

I removed it in favour of:
1. Using `waitPort` in the function that starts the framework server: https://github.com/netlify/cli/pull/1196/files#diff-8a2f65e3aede3ea9696e546766f50f03R82
2. When starting the functions server - wait on a promise until the server is ready: https://github.com/netlify/cli/pull/1196/files#diff-cf2bd94d7688f0711bbd931e6c5e0397R373